File: [E]heap/[E]heap/973-k-closest-points-to-origin.py
from heapq import *
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        heap = []
        for x, y in points:
            distance = (x**2 + y**2) ** 0.5
            heappush(heap, (- distance, x, y))
            if len(heap) > k:
                heappop(heap)
                
        return [[x, y] for _, x, y in heap]

# time O(nlogk)
# space O(k)
# using heap and top k problem (based on heap) and max heap

File: [E]heap/[E]heap/973-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> Integer.compare(b[0], a[0]));
        int[][] res = new int[k][2];
        for (int[] p: points) {
            heap.offer(new int[]{p[0] * p[0] + p[1] * p[1], p[0], p[1]});
            if (heap.size() > k) {
                heap.poll();
            }
        }
        while (!heap.isEmpty()) {
            int idx = heap.size() - 1;
            int[] p = heap.poll();
            res[idx][0] = p[1];
            res[idx][1] = p[2];
        }
        return res;
    }
}

// time O(nlogk)
// space O(k)
// using heap and top k problem (based on heap) and max heap